scylla-cdc/src/e2e_tests.rs

```diff
@@ -0,0 +1,463 @@
+                .execute(&self.update_query, Test::get_value_list(&pk_vec, ck, v))
+                .await?;
+            let operation = Operation::new(OperationType::RowUpdate, Some(ck), v);
+            self.push_back(pk_vec, operation);
+
+            Ok(())
+        }
+
+        async fn compare(&mut self, result: OperationsMap) -> bool {
+            let mut is_ok = true;
+
+            for (pk, actual_operations) in result.lock().await.iter_mut() {
+                let expected_operations = match self.performed_operations.get_mut(pk) {
+                    Some(ops) => ops,
+                    None => {
+                        eprintln!("Unexpected primary key {:?}", pk);
+                        is_ok = false;
+                        continue;
+                    }
+                };
+                let mut i = 0;
+
+                while let Some(next_expected) = expected_operations.pop_front() {
+                    i += 1;
+                    let next_actual = match actual_operations.pop_front() {
+                        Some(op) => op,
+                        None => {
+                            eprintln!("Too little read operations for primary key {:?}. Missing operations count: {}", pk, expected_operations.len() + 1);
+                            is_ok = false;
+                            break;
+                        }
+                    };
+
+                    if next_expected != next_actual {
+                        eprintln!("Operation no. {} not matching for primary key {:?}.", i, pk);
+                        eprintln!("\tExpected: {:?}, actual: {:?}", next_expected, next_actual);
+                        is_ok = false;
+                        break;
+                    }
+                }
+
+                if is_ok && expected_operations.len() != actual_operations.len() {
+                    eprintln!(
+                        "Too many read operations for primary key {:?}. Operations left: {}",
+                        pk,
+                        actual_operations.len()
+                    );
+                    is_ok = false;
+                }
+            }
+
+            is_ok
+        }
+
+        async fn test_cdc(mut self, start: chrono::Duration) -> Result<()> {
+            let results = Arc::new(Mutex::new(HashMap::new()));
+            let factory = Arc::new(TestConsumerFactory::new(Arc::clone(&results)));
+            let end = now();
+
+            let (mut tester, _handle) = CDCLogReader::new(
+                Arc::clone(&self.session),
+                "e2e_test".to_string(),
+                self.table_name.clone(),
+                start,
+                end,
+                time::Duration::from_millis(WINDOW_SIZE),
+                time::Duration::from_millis(SAFETY_INTERVAL),
+                time::Duration::from_millis(SLEEP_INTERVAL),
+                factory,
+            );
+
+            sleep(time::Duration::from_secs(15)).await;
+            tester.stop();
```

Comment:
This causes the test to depend on timing. You can't be sure that everything completes within 15ms. Moreover, after calling `stop()` the log reader worker will stop ASAP but not immediately.

As you already provided an end timestamp in the constructor, it will stop itself after reaching it. In order to wait until the worker actually finishes, you need to `.await` the handle returned from `CDCLogReader::new`.

Reply:
I'm not sure that works. The tests don't come to an end, I even got messages like this one: `test e2e_tests::tests::e2e_test_int_pk has been running for over 60 seconds`. 
I replaced these two lines with this one: 
```
handle.await.unwrap();
```

Reply:
In that case, it sounds like we might have a bug in the log reader and it should be fixed. cc: @Ponewor

Reply:
It looks like the CDCLogReader, after creating StreamReaders, should set their upper timestamps to self.end_timestamp, but it doesn't.

Reply:
That might be it, I just did some trial and error attempts and it seems that the problem is with the default end timestamp - it works fine when we use `stop` or even `stop_at`. I'm going to try it.

Reply:
Yup, it works.